Tests/Remora.Rest.Core.Tests/Optional/OptionalTests.cs

```diff
@@ -584,4 +584,208 @@ public static void ReturnsContainedCombinedWithTrueForContainedValue()
             Assert.Equal(HashCode.Combine(1, true), a.GetHashCode());
         }
     }
+
+    public static class OrDefault
+    {
+        [Fact]
+        public static void ReturnsDefaultValueOfTypeForStructs()
+        {
+            var a = default(Optional<int>);
+            Assert.Equal(default, a.OrDefault());
+        }
+
+        [Fact]
+        public static void ReturnsNullValueForClasses()
+        {
+            var a = default(Optional<object>);
+            Assert.Equal(null, a.OrDefault());
+        }
+
+        [Fact]
+        public static void ReturnsExistingValueIfContainsValue()
+        {
+            Optional<int> a = 1;
+            Assert.Equal(1, a.OrDefault());
+            Assert.Equal(1, a.OrDefault(2));
+        }
+
+        [Fact]
+        public static void ReturnsProvidedValueIfNull()
+        {
+            Optional<string?> a = null;
+            Assert.Equal("Expected", a.OrDefault("Expected"));
+
+            Optional<int?> b = null;
+            Assert.Equal(1, b.OrDefault(1));
+        }
+
+        [Fact]
+        public static void ReturnsProvidedDefaultValueIfDoesNotContainValue()
+        {
+            var a = default(Optional<string>);
+            Assert.Equal("Expected", a.OrDefault("Expected"));
+
+            var b = default(Optional<int>);
+            Assert.Equal(1, b.OrDefault(1));
+        }
+    }
+
+    public static class OrThrow
+    {
+        [Fact]
+        public static void ThrowsIfDoesNotContainValue()
+        {
+            var a = default(Optional<int>);
+            Assert.Throws<InvalidOperationException>(()
+                => a.OrThrow(static () => new InvalidOperationException("Expected")));
+        }
+
+        [Fact]
+        public static void DoesNotThrowIfContainsValue()
+        {
+            Optional<int> a = 1;
+
+            var exception =
+                Record.Exception(() => a.OrThrow(static () => new InvalidOperationException("Not expected")));
+            Assert.Null(exception);
+        }
+    }
+
+    public static class AsNullableOptional
+    {
+        [Fact]
+        public static void ProducesCorrectValueWhenEmpty()
```

Comment:
Doing this with a `struct` type might be useful too, since that's actually going to have a different type instead of just different nullability annotations.

Reply:
This function can't produce correct results for structs, compiler limitation

Reply:
(Specifically, if you call it you just get an Optional of the same type.)

Reply:
It's possible to do this by a little bit of hackery - move the method out into an extension class, then have one extension method class for structs and one for classes. That way, you can have overloads that only differ in constraints.